ledger/acctupdates_test.go

```diff
@@ -1637,3 +1637,256 @@ func TestCachesInitialization(t *testing.T) {
 	require.Equal(t, int(proto.MaxBalLookback), len(au.deltas))
 	require.Equal(t, recoveredLedgerRound-basics.Round(proto.MaxBalLookback), au.dbRound)
 }
+
+// TestSplittingConsensusVersionCommits tests the a sequence of commits that spans over multiple consensus versions works correctly.
+func TestSplittingConsensusVersionCommits(t *testing.T) {
+	initProtocolVersion := protocol.ConsensusV20
+	initialProtoParams := config.Consensus[initProtocolVersion]
+
+	initialRounds := uint64(1)
+
+	ml := makeMockLedgerForTracker(t, true, int(initialRounds), initProtocolVersion)
+	ml.log.SetLevel(logging.Warn)
+	defer ml.Close()
+
+	accountsCount := 5
+	accts := []map[basics.Address]basics.AccountData{randomAccounts(accountsCount, true)}
+	rewardsLevels := []uint64{0}
+
+	pooldata := basics.AccountData{}
+	pooldata.MicroAlgos.Raw = 1000 * 1000 * 1000 * 1000 * 1000 * 1000
+	pooldata.Status = basics.NotParticipating
+	accts[0][testPoolAddr] = pooldata
+
+	sinkdata := basics.AccountData{}
+	sinkdata.MicroAlgos.Raw = 1000 * 1000 * 1000 * 1000
+	sinkdata.Status = basics.NotParticipating
+	accts[0][testSinkAddr] = sinkdata
+
+	au := &accountUpdates{}
+	au.initialize(config.GetDefaultLocal(), ".", initialProtoParams, accts[0])
+	err := au.loadFromDisk(ml)
+	require.NoError(t, err)
+	defer au.close()
+
+	// cover initialRounds genesis blocks
+	rewardLevel := uint64(0)
+	for i := 1; i < int(initialRounds); i++ {
+		accts = append(accts, accts[0])
+		rewardsLevels = append(rewardsLevels, rewardLevel)
+	}
+
+	extraRounds := uint64(39)
+
+	// write the extraRounds rounds so that we will fill up the queue.
+	for i := basics.Round(initialRounds); i < basics.Round(initialRounds+extraRounds); i++ {
+		rewardLevelDelta := crypto.RandUint64() % 5
+		rewardLevel += rewardLevelDelta
+		accountChanges := 2
+
+		updates, totals := randomDeltasBalanced(accountChanges, accts[i-1], rewardLevel)
+		prevTotals, err := au.Totals(basics.Round(i - 1))
+		require.NoError(t, err)
+
+		newPool := totals[testPoolAddr]
+		newPool.MicroAlgos.Raw -= prevTotals.RewardUnits() * rewardLevelDelta
+		updates.Upsert(testPoolAddr, newPool)
+		totals[testPoolAddr] = newPool
+
+		blk := bookkeeping.Block{
+			BlockHeader: bookkeeping.BlockHeader{
+				Round: basics.Round(i),
+			},
+		}
+		blk.RewardsLevel = rewardLevel
+		blk.CurrentProtocol = initProtocolVersion
+
+		delta := ledgercore.MakeStateDelta(&blk.BlockHeader, 0, updates.Len(), 0)
+		delta.Accts.MergeAccounts(updates)
+		ml.addMockBlock(blockEntry{block: blk}, delta)
+		au.newBlock(blk, delta)
+		accts = append(accts, totals)
+		rewardsLevels = append(rewardsLevels, rewardLevel)
+	}
+
+	newVersionBlocksCount := uint64(47)
+	newVersion := protocol.ConsensusV21
+	// add 47 more rounds that contains blocks using a newer consensus version, and stuff it with MaxBalLookback
+	for i := basics.Round(initialRounds + extraRounds); i < basics.Round(initialRounds+initialProtoParams.MaxBalLookback+extraRounds+newVersionBlocksCount); i++ {
```

Comment:
move `basics.Round(initialRounds+initialProtoParams.MaxBalLookback+extraRounds+newVersionBlocksCount` into a separate var and re-use below in `committedUpTo` to make it obvious what happens.